router/throttler/adaptiveRateLimit.go

```diff
@@ -0,0 +1,112 @@
+package throttler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+)
+
+type timers struct {
+	timer        time.Duration
+	limitReached map[string]bool
+	mu           sync.Mutex
+	limitSet     bool
+}
+
+func SetupRouterAdaptiveRateLimiter(ctx context.Context, errorCh <-chan string) func(destName, destID string, limit int64) int64 {
+	shortTimeFrequency := config.GetDuration("Router.throttler.adaptiveRateLimit.shortTimeFrequency", 5, time.Second)
+	longTimeFrequency := config.GetDuration("Router.throttler.adaptiveRateLimit.longTimeFrequency", 15, time.Second)
+
+	shortTimer := &timers{
+		timer:        shortTimeFrequency,
+		limitReached: make(map[string]bool),
+	}
+	longTimer := &timers{
+		timer:        longTimeFrequency,
+		limitReached: make(map[string]bool),
+	}
+
+	go shortTimer.runLoop(ctx)
+	go longTimer.runLoop(ctx)
+
+	go func() {
```

Comment:
[blocking]
How do we wait for the go routines to stop? Also, currently, in your implementation, you are passing context.Background()

```go
		f.adaptiveRateLimiter = SetupRouterAdaptiveRateLimiter(context.Background(), f.limitReachedPerDestination)
```

So the go routines will never stop